src/Filters/remez_fir.jl

```diff
@@ -0,0 +1,845 @@
+    The number of taps is the number of terms in the filter, or the filter 
+    order plus one.
+- `bands::Array`: A monotonic sequence containing the band edges in Hz.
+    All elements must be non-negative and less than half the sampling
+    frequency as given by `Hz`.
+- `desired::Array`:A sequence half the size of bands containing the desired 
+    gain in each of the specified bands.
+- `weight::Array`: (optional)
+    A relative weighting to give to each band region. The length of
+    `weight` has to be half the length of `bands`.
+- `Hz::Real`: The sampling frequency in Hz. Default is 1.
+- `filter_type::String`: {'bandpass', 'differentiator', 'hilbert'}, optional
+    The type of filter:
+      'bandpass' : flat response in bands. This is the default.
+      'differentiator' : frequency proportional response in bands.
+      'hilbert' : filter with odd symmetry, that is, type III
+                  (for even order) or type IV (for odd order)
+                  linear phase filters.
+- `maxiter::Integer`: (optional)
+    Maximum number of iterations of the algorithm. Default is 25.
+- `grid_density:Integer`: (optional)
+    Grid density. The dense grid used in `remez` is of size
+    ``(numtaps + 1) * grid_density``. Default is 16.
+
+# Returns
+- `h::Array{Float64,1}`: A rank-1 array containing the coefficients of the optimal
+    (in a minimax sense) filter.
+
+[^McClellan1973a]: 
+J. H. McClellan and T. W. Parks, A unified approach to the
+design of optimum FIR linear phase digital filters,
+IEEE Trans. Circuit Theory, vol. CT-20, pp. 697-701, 1973.
+
+[^McClellan1973b]: 
+J. H. McClellan, T. W. Parks and L. R. Rabiner, A Computer
+Program for Designing Optimum FIR Linear Phase Digital
+Filters, IEEE Trans. Audio Electroacoust., vol. AU-21,
+pp. 506-525, 1973.
+
+# Examples
+Construct a length 35 filter with a passband at 0.15-0.4 Hz 
+(desired response of 1), and stop bands at 0-0.1 Hz and 0.45-0.5 Hz
+(desired response of 0). Note: the behavior in the frequency ranges between 
+those bands - the transition bands - is unspecified.
+
+```julia-repl
+julia> bpass = remez(35, [0 0.1 0.15 0.4 0.45 0.5], [0 1 0])
+```
+
+You can trade-off maximum error achieved for transition bandwidth. 
+The wider the transition bands, the lower the maximum error in the
+bands specified. Here is a bandpass filter with the same passband, but
+wider transition bands.
+
+```julia-repl
+julia> bpass2 = remez(35, [0 0.08 0.15 0.4 0.47 0.5], [0 1 0])
+```
+
+Here we compute the frequency responses and plot them in dB.
+
+```julia-repl
+using PyPlot
+b = DSP.Filters.PolynomialRatio(bpass, [1.0]);
+b2 = DSP.Filters.PolynomialRatio(bpass2, [1.0]);
+f = linspace(0, 0.5, 1000)
+plot(f, 20*log10.(abs.(freqz(b,f,1.0))))
+plot(f, 20*log10.(abs.(freqz(b2,f,1.0))))
+grid()
+```
+"""
+function remez(numtaps::Integer, bands::Array, desired::Array; 
+               weight::Array=[], 
+               Hz::Real=1.0, 
+               filter_type::String="bandpass",
```

Comment:
`filter_type` should a `Symbol`, or an `@enum FilterType bandpass=1 differentiator=2 hilbert=3`. Then comparison is fast so that we can eliminate the `jtype` variable.